src/actions/clipboard.ts

```diff
@@ -0,0 +1,368 @@
+  }
+
+  /**
+   * Uninstall this action as both a keyboard shortcut and a context menu item.
+   * Reinstall the original context menu action if possible.
+   */
+  uninstall() {
+    ContextMenuRegistry.registry.unregister('blockCutFromContextMenu');
+    ContextMenuRegistry.registry.unregister('blockCopyFromContextMenu');
+    ContextMenuRegistry.registry.unregister('blockPasteFromContextMenu');
+
+    ShortcutRegistry.registry.unregister(Constants.SHORTCUT_NAMES.CUT);
+    ShortcutRegistry.registry.unregister(Constants.SHORTCUT_NAMES.COPY);
+    ShortcutRegistry.registry.unregister(Constants.SHORTCUT_NAMES.PASTE);
+  }
+
+  /**
+   * Create and register the keyboard shortcut for the copy action.
+   */
+  private registerCopyShortcut() {
+    const copyShortcut: ShortcutRegistry.KeyboardShortcut = {
+      name: Constants.SHORTCUT_NAMES.COPY,
+      preconditionFn: this.copyPrecondition.bind(this),
+      callback: this.copyCallback.bind(this),
+      keyCodes: [
+        createSerializedKey(KeyCodes.C, [KeyCodes.CTRL]),
+        createSerializedKey(KeyCodes.C, [KeyCodes.ALT]),
+        createSerializedKey(KeyCodes.C, [KeyCodes.META]),
+      ],
+      allowCollision: true,
+    };
+    ShortcutRegistry.registry.register(copyShortcut);
+  }
+
+  /**
+   * Register the copy block action as a context menu item on blocks.
+   * This function mixes together the keyboard and context menu preconditions
+   * but only calls the keyboard callback.
+   */
+  private registerCopyContextMenuAction() {
+    const copyAction: ContextMenuRegistry.RegistryItem = {
+      displayText: (scope) => `Copy (${this.getPlatformPrefix()}C)`,
+      preconditionFn: (scope) => {
+        const ws = scope.block?.workspace;
+        if (!ws) return 'hidden';
+
+        return this.copyPrecondition(ws) ? 'enabled' : 'disabled';
+      },
+      callback: (scope) => {
+        const ws = scope.block?.workspace;
+        if (!ws) return;
+        return this.copyCallback(ws);
+      },
+      scopeType: ContextMenuRegistry.ScopeType.BLOCK,
+      id: 'blockCopyFromContextMenu',
+      weight: BASE_WEIGHT + 1,
+    };
+
+    ContextMenuRegistry.registry.register(copyAction);
+  }
+
+  /**
+   * Precondition function for copying a block from keyboard
+   * navigation. This precondition is shared between keyboard shortcuts
+   * and context menu items.
+   *
+   * @param workspace The `WorkspaceSvg` where the shortcut was
+   *     invoked.
+   * @returns True iff `copyCallback` function should be called.
+   */
+  private copyPrecondition(workspace: WorkspaceSvg) {
+    if (!this.canCurrentlyEdit(workspace)) return false;
+    switch (this.navigation.getState(workspace)) {
+      case Constants.STATE.WORKSPACE:
+        const curNode = workspace?.getCursor()?.getCurNode();
+        const source = curNode?.getSourceBlock();
+        return !!(
+          source?.isDeletable() &&
+          source?.isMovable() &&
+          !Gesture.inProgress()
```

Comment:
This and similar lines cause https://github.com/google/blockly-keyboard-experimentation/issues/237

Filed as a separate bug so that the structural change can land.